File: src/cmcandy/C_Language_Answers/_0328.c

#include <stdio.h>
#include <stdlib.h>
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode
{
    int val;
    struct ListNode *next;
};
struct ListNode *oddEvenList(struct ListNode *head)
{
    //奇偶两条数组，最后合起来
    struct ListNode *cur = head;
    struct ListNode *oddNum, *oddHead;
    struct ListNode *evenNum, *evenHead;

    oddHead = (struct ListNode *)malloc(sizeof(struct ListNode));
    evenHead = (struct ListNode *)malloc(sizeof(struct ListNode));
    oddNum = oddHead;
    evenNum = evenHead;
    int tag = 0;
    while (cur != NULL)
    {
        //if old
        if (tag == 0)
        {
            oddNum->next = cur;
            oddNum = oddNum->next;
            tag = 1;
        }
        else if (tag == 1)
        {
            //if even
            evenNum->next = cur;
            evenNum = evenNum->next;
            tag = 0;
        }
        cur = cur->next;
    }
    evenNum->next = NULL;
    //mix
    oddNum->next = evenHead->next;
    return oddHead->next;
}

File: src/cmcandy/java_language_answers/_0328.java
public class _0328 {
    /**
     * Definition for singly-linked list. public class ListNode { int val; ListNode
     * next; ListNode() {} ListNode(int val) { this.val = val; } ListNode(int val,
     * ListNode next) { this.val = val; this.next = next; } }
     */

    static public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    static class Solution {
        public ListNode oddEvenList(ListNode head) {

            int tag = 0;
            ListNode oddLis = new ListNode();
            ListNode evenLis = new ListNode();
            ListNode curOdd = oddLis;
            ListNode curEven = evenLis;

            ListNode cur = head;

            while (cur != null) {
                if (tag == 0) {
                    curOdd.next = cur;
                    curOdd = curOdd.next;
                    tag = 1;
                } else if (tag == 1) {
                    curEven.next = cur;
                    curEven = curEven.next;
                    tag = 0;
                }
                cur = cur.next;
            }
            curEven.next = null;
            curOdd.next = evenLis.next;
            return oddLis.next;

        }
    }
    public static void main(String[] args) {
        Solution solution = new Solution();
        ListNode head2 = new ListNode(1);
        ListNode head = head2;
        ListNode cur = new ListNode(2);
        head.next = cur;
        head = head.next;
        cur = new ListNode(3);
        head.next = cur;
        head = head.next;
        cur = new ListNode(4);
        head.next = cur;
        head = head.next;
        cur = new ListNode(5);
        head.next = cur;
        head = head.next;
        System.out.println(solution.oddEvenList(head2));
    }
}